File: src/main/kotlin/org/araqnid/typevariance/kotlin/EventDeserializer.kt
package org.araqnid.typevariance.kotlin

import org.araqnid.typevariance.BasicEvent
import org.araqnid.typevariance.Event

interface EventDeserializer<out E> {
    fun deserialize(bytes: ByteArray): E
}

val eventDeserializer: EventDeserializer<Event> = object : EventDeserializer<Event> {
    override fun deserialize(bytes: ByteArray): Event = TODO()
}

val basicEventDeserializer: EventDeserializer<Event> = object : EventDeserializer<BasicEvent> {
    override fun deserialize(bytes: ByteArray): BasicEvent = TODO()
}

// val invalid: EventDeserializer<BasicEvent> = eventDeserializer

File: src/main/scala/org/araqnid/typevariance/scala/EventDeserializer.scala
package org.araqnid.typevariance.scala

import org.araqnid.typevariance.{BasicEvent, Event}

trait EventDeserializer[+E] {
  def deserialize(bytes: Array[Byte]): E
}

object EventDeserializers {
  val eventDeserializer: EventDeserializer[Event] = new EventDeserializer[Event] {
    override def deserialize(bytes: Array[Byte]): Event = ???
  }

  val basicEventDeserializer: EventDeserializer[Event] = new EventDeserializer[BasicEvent] {
    override def deserialize(bytes: Array[Byte]): BasicEvent = ???
  }

  // val invalid: EventDeserializer[BasicEvent] = eventDeserializer
}
